PHP 8 attribute support

themasch opened the pull request:
**Disclaimer first:** It may be the case, that by some magic unknown to me, this library already support accessing the attribute based doctrine metadata, and I was just not smart enough to get it to work. If thats the case, I am sorry for wasting your time.

This should allow users to use this with the Doctrine AttributeDriver for metadata, instead of using the AnnotationDriver. This requires the that all `GraphQL\Doctrine\Annotation`s are written as annotations as well, though.
The alternative would be to use two readers, one that reads annotations and one that reads attributes, and fall back to one of the if the other does not yield a result.

There are a few quirks in this version that I hope to fix in the next commits, but its getting late... 

First, the IDE complaints if we use the named arguments in attributes, that do not have constructor parameters with that name. Makes sense, I guess its Doctrine magic that makes that work at the moment, but we may not want to rely on that. As far as I can see there are two possible routes to fix that: 
Sacrefice the nice syntax of `#[API\Field(args: [new API\Argument(name: 'status', type: '?string')])]`
 and go with `#[API\Field(args: [new API\Argument(['name' => 'status', 'type' => '?string'])])]`, which uses the default constructor and seems to work fine. But IMHO its ugly, it does not support IDE type hinting, and we'd also give up type checking.
Not my prefered option. 

The other way would be, to give all annotations a constructor that works for them, add the `@NamedArgumentConstructor` annotation, and everything should be fine. This worked for most of them, but `Field` gave me some pain. I did not went into that rabbit hole (yet) but it seems that the type hints on that class lie and it may also be re-used to hold some other DTOs? 

More research is required. 

Also, the new test are kind of messy. I did not want to restructure to much, but I also wanted to have at least some basic coverage for the first step. Thats why the current state is just some copy & paste with a coat of sugar. This needs to be improved. 

Lastly, the AttributeReader we wrap here to pretent that its an Annotation\Reader is marked as `@Internal` by doctrine, so usign it directly does not seem advisable. Not sure if we have a good alterantive over this, short of copy-pasting it.

PowerKiKi commented:
This will probably take a bit of time until we can review, merge and release this, but it's definitely the direction we want to go. So thank you for starting this.

Here are a few quick answers to guide you, without even reading your code yet:

> this library already support accessing the attribute based doctrine

no, it doesn't. So this PR is very welcome.

> requires the that all GraphQL\Doctrine\Annotations are written as annotations as well

This is exactly what I planned. Annotation must be dropped entirely in favor of attributes.

Are you aware of a tool to migrate annotation to attribute ? specific for Doctrine ? or generic ?

> The alternative would be to use two readers

No. Assume the breaking change, and keep it simple with a single reader.

> Sacrifice the nice syntax [...] give up type checking

No thanks. Type checking must be preserved.

> give all annotations a constructor that works for them

Yes, please !

> but Field gave me some pain. I did not went into that rabbit hole (yet) but it seems that the type hints on that class lie and it may also be re-used to hold some other DTOs?

Totally correct. We do use `Field` as a kind of DTO. It's not great how it is now. But that was the most reasonable solution I found back then. If you see a way to do it differently, feel free to do so.

> marked as @Internal by doctrine

I am willing to take the risk to use internal things if necessary, as long as they are well covered in our unit tests.

themasch committed:
Initial draft of PHP 8 attribute support.

This should allow users to use this with the Doctrine AttributeDriver for metadata, instead of using the AnnotationDriver. This requires the that all `GraphQL\Doctrine\Annotation`s are written as annotations as well, though.
The alternative would be to use two readers, one that reads annotations and one that reads attributes, and fall back to one of the if the other does not yield a result.

themasch commented:
Sorry about the force pushes, just noticed that the wrong email adress was used in the commits, changed that.

themasch commented:
Okay, small update: 
seems like the `@NamedArgumentConstructor` works fine for everything beside `Input` and `Argument`. Not sure what trouble `Field` gave me last time I check but it seems to behave itself well this time. 
The problem ist that with the current approach, there is not way to differentiate between `$defaultValue = null` and `no default value is set`, because, with the defaultValue beeing a constructor argument, it may be nullable, but we have to pick an inital value, and if we pick `null` then everything that has no defaultValue set will be null by default. 

I found one hacky way around that: predefine a well-known constant value, use that as the initial/default value for the `$defaultValue` parameter, and use that to identifiy if `null` was given or not:

```php
final class Input extends AbstractAnnotation
{
    private const NO_VALUE_PASSED = '_hacky_no_value_past_find_a_better_solution_for_this_';

    public function __construct(/** omitted **/mixed $defaultValue = self::NO_VALUE_PASSED) 
    {
        $settings = [ /** the other stuff **/  ];
        if ($defaultValue !== self::NO_VALUE_PASSED) {
            $settings['defaultValue'] = $defaultValue;
        }

        parent::__construct($settings);
    }
}
```

To be clear: I really do not like this solution, but I am also not smart enough to come up with a better one right now.

An alternative could be inventing a special value to use instead of `null` for when you really want `null`, leading to "`defaultValue = null` means no default value, `defaultValue = Input::NULL`  means `null`" or something of that kind. 
That would break BC and be hard to communicate to users, I think. So even worse?

¯\_(ツ)_/¯

themasch committed:
Add the last missing named argument constructors.

This currently uses an ugly, dirty hack to work around a missing `undefined` value.

PowerKiKi commented:
```php
private const NO_VALUE_PASSED
```

seems like an acceptable workaround to me because it is private and thus never accessible to anybody at all. It is a very localized workaround that cannot leak anywhere. And as I also am not smart enough to come up with an alternative, I'd say let's go for that.

driehle commented:
@themasch @PowerKiKi 

> I found one hacky way around that: predefine a well-known constant value, use that as the initial/default value for the `$defaultValue` parameter, and use that to identifiy if `null` was given or not:

To me, this looks like you should simply omit all parameters from the constructor and use [func_num_args()](https://www.php.net/manual/en/function.func-num-args.php) to check if any parameters have been passed and [func_get_args()](https://www.php.net/manual/en/function.func-get-args.php) to actually access them.

PowerKiKi commented:
I guess that would work too. But I actually prefer the more explicit private const suggested.

I *will* review and merge this PR some days, but meanwhile thanks for your patience though...

driehle commented:
You can keep the explicit notation and still use `func_num_args()`, see https://3v4l.org/kMHUp for an example.

PowerKiKi commented:
This been a long time in the making, but we're finally migrating to attributes.

Thank you @themasch for starting this work. It was very helpful :tada:
